Mathlib/Combinatorics/SimpleGraph/ConcreteColorings.lean

```diff
@@ -48,4 +48,28 @@ theorem chromaticNumber_pathGraph (n : ℕ) (h : 2 ≤ n) :
   · simpa only [pathGraph_two_eq_top, chromaticNumber_top] using
       chromaticNumber_mono_of_embedding (pathGraph_two_embedding n h)
 
+theorem Coloring.bicoloring_apply_iff_of_walk {α} {G : SimpleGraph α}
+    (c : G.Coloring Bool) {u v : α} (p : G.Walk u v) :
+    c v ↔ (c u ↔ Even p.length) := by
```

Comment:
Maybe
```suggestion
theorem Coloring.even_length_iff_congr {α} {G : SimpleGraph α}
    (c : G.Coloring Bool) {u v : α} (p : G.Walk u v) :
    Even p.length ↔ (c u ↔ c v) := by
```
where you can use the (apparently missing!) lemmas `iff_assoc : (p ↔ (q ↔ r)) ↔ ((p ↔ q) ↔ r)`, `iff_left_comm : (p ↔ (q ↔ r)) ↔ (q ↔ (p ↔ r))` and `iff_assoc : ((p ↔ q) ↔ r) ↔ ((p ↔ r) ↔ q)` and the existing `iff_comm` to get back to your original statement

Reply:
I think `Coloring.bicoloring_apply_iff_of_walk` is the more useful form. The point of this is that you can rewrite `c v` to `c u` or `not (c u)` conditionally on whether or not `Even p.length` is true.

The `Coloring.even_length_iff_congr` form is nice and symmetric though, so it could be worth having around too.

Reply:
I suspect what we actually want is `Even p.length ↔ (c u ↔ c v)` and `Odd p.length ↔ (c u ↔ ¬ c v)`(/`Odd p.length ↔ (¬ c u ↔ c v)` so that we really "can rewrite `c v` to `c u` or `not (c u)` conditionally on whether or not `Even p.length` is true"

Reply:
@YaelDillies The benefit of the `Coloring.bicoloring_apply_iff_of_walk` formulation is that it introduces the `Even p.length` term. I think many times you won't have `Even p.length` or `Odd p.length` in context or in an easy-to-express form. It's using `↔` more like a boolean operator to conditionally negate.

Reply:
I guess I am missing the greater context here

Reply:
Hmm... the fact that these are Bool-valued rather than Prop-valued walks makes the use case I have in mind not work very well. The statement would need to be `c u = (c v == decide (Even p.length))` instead.

I think in many cases, you would also want the forms `Even p.length -> c u = c v` and `Odd p.length -> c u = ! c v`, since that way the `Even`/`Odd` conditions can create new goals. The purpose is to be able to use a walk to rewrite `c u` to `c v`.

Reply:
I added the `Even p.length ↔ (c u ↔ c v)` and `Odd p.length ↔ (¬c u ↔ c v)` versions
I also added missing `iff_assoc`, `iff_left_comm` ans some more (witch may need better names) for probing the other theorems from one. I added them in `Mathlib/Init/Logic.lean`, but maybe you think that they should go somewhere else